File: 의진/프로그래머스/매칭점수.py
import re


def solution(word, pages):
    url2idx = {}
    move_data = [[[], []] for _ in range(len(pages))]
    scores = [0] * len(pages)

    for i, page in enumerate(pages):
        # 1. 페이지 주소 가져오기
        # page_url = re.search('<meta property="og:url" content="(\S+)"', page).group(1) -> 정규표현식 버전
        # page_url = page.split('"og:url" content="')[1].split('"/>')[0] -> 틀림 : "og.url"이라는게 다른 곳에 있으면 털림
        page_url = page.split('<meta property="og:url" content="')[1].split('"/>')[0]
        url2idx[page_url] = i

    for i, page in enumerate(pages):
        # 2. body 부분 추출
        body = page.split("<body>")[1].split("</body>")[0]

        # 3. 링크 추출
        # exiosLink = re.findall('<a href="(.+)">', '<a href="https://rerfdb"> test="dfdqom">') -> 뒤에 >가 있으면 틀림
        # exiosLink = re.findall('<a href="(https://[\S]*)"', page) -> 이렇게 해야함
        links = list(map(lambda x: x.split('">')[0], body.split('<a href="')[1:]))
        # links = body.split('<a href="')[1:]

        # 4. link 반영
        for link in links:
            if url2idx.get(link) is not None:
                move_data[i][1].append(
                    url2idx[link]
                )  # [1] = 이 페이지에서 어디 페이지로 가냐
                move_data[url2idx[link]][0].append(i)  # [0] = 이 페이지로 오는 페이지들
            else:
                move_data[i][1].append("x")

        # # 5. text 추출
        # texts = list(
        #     map(lambda x: x.split("<a href")[0].lower().strip(), body.split("</a>"))
        # )

        # 6. 기본점수 계산
        for w in re.findall("[a-z]+", page.lower()):  # 페이지 전체에서 단어검색
            if w == word.lower():
                scores[i] += 1

    # print(scores)
    cur_max_score = 0
    cur_max_i = 0
    for i in range(len(pages)):
        final_score = 0
        final_score += scores[i]

        for adjacent in move_data[i][0]:
            final_score += scores[adjacent] / len(move_data[adjacent][1])

        if cur_max_score < final_score:
            cur_max_score = final_score
            cur_max_i = i

    return cur_max_i

File: 명준/매칭점수.kt
//package com.example.myapplication
//
//
//
///*
//한 웹페이지에 대해서 기본점수, 외부 링크 수, 링크점수, 그리고 매칭점수를 구할 수 있다.
//기본점수 = 텍스트 중 검색어가 등장하는 횟수
//외부링크 = (해당웹페이지) 에서 (다른 외부 페이지) 로 연결된 링크의 갯수
//링크점수 = (해당 페이지로 링크가 걸린 다른 웹페이지)의 기본점수 / 링크점수   의
//매칭점수 = 기본점수 + 링크점수
//
//1.HTML추출하고 파싱하기
//URL , 외부링크 ,    body에 등장한 검색횟수 구하기
//123 다 <전달할> 점수 계산한 후 받은 링크점수 계산하기
//매칭점수 + 링크점수로 매칭점수 계산후
//비교하여 index 반환, 같다면 낮은거
//
//
//
//
//
//검색어 words와 웹 페이지의 HTML 목록 pages가 주어질때 , 매칭점수가 가장 높은 웹 페이지의 index를 구하라, 여러개라면 번호가 작은거
//
//
//제한사항
//
//pages는 HTML 형식의 웹페이지가 문자열 형태로 들어있는 배열이고, 길이는 1 이상 20 이하이다.
//한 웹페이지 문자열의 길이는 1 이상 1,500 이하이다.
//웹페이지의 index는 pages 배열의 index와 같으며 0부터 시작한다.
//한 웹페이지의 url은 HTML의 태그 내에 태그의 값으로 주어진다.
//
//파싱할때
//
//
//입력 예시
//
//<html lang="ko" xml:lang="ko" xmlns="http://www.w3.org/1999/xhtml">
//<head>
//  <meta charset="utf-8">
//  <meta property="og:url" content="https://a.com"/>
//</head>
//<body>
//Blind Lorem Blind ipsum dolor Blind test sit amet, consectetur adipiscing elit.
//<a href="https://b.com"> Link to b </a>
//</body>
//</html>
//
//<html lang="ko" xml:lang="ko" xmlns="http://www.w3.org/1999/xhtml">
//<head>
//  <meta charset="utf-8">
//  <meta property="og:url" content="https://b.com"/>
//</head>
//<body>
//Suspendisse potenti. Vivamus venenatis tellus non turpis bibendum,
//<a href="https://a.com"> Link to a </a>
//blind sed congue urna varius. Suspendisse feugiat nisl ligula, quis malesuada felis hendrerit ut.
//<a href="https://c.com"> Link to c </a>
//</body>
//</html>
//
//
//<html lang="ko" xml:lang="ko" xmlns="http://www.w3.org/1999/xhtml">
//<head>
//  <meta charset="utf-8">
//  <meta property="og:url" content="https://c.com"/>
//</head>
//
//<body>
//Ut condimentum urna at felis sodales rutrum. Sed dapibus cursus diam, non interdum nulla tempor nec. Phasellus rutrum enim at orci consectetu blind
//
//<a href="https://a.com"> Link to a </a>
//</body>
//
//</html>
//
//
//1.HTML추출하고 파싱하기 -> URL , 외부링크 , 등장한 words 횟수구하기
//
// 다 <전달할> 점수 계산한 후 받은 링크점수 계산하기
//매칭점수 + 링크점수로 매칭점수 계산후
//비교하여 index 반환, 같다면 낮은거
//
//
//regex .*?
//
// */
//
//data class Pages(
//    val index : Int,
//    val url : String,
//    val basicScore : Int,
//    val link : List<String>,
//    var matchingScore : Double = 0.0 ,
//    var totalScore: Double = 0.0
//)
//
//fun findUrl(page : String) : String{
//    val urlRegex = ("<meta property=\"og:url\" content=\"(.*?)\"").toRegex()
//    return urlRegex.find(page)!!.groupValues[1]
//}
//
//fun findLinks(page: String) : List<String>{
//    val linkRegex = "<a href=\"(.*?)\"".toRegex()
//    return linkRegex.findAll(page).map { it.groupValues[1] }.toList()
//}
//
//
//fun wordsCount(page: String, word: String) : Int{
//    val body = page.substringAfter("<body>").substringBefore("</body>").lowercase()
//    val pattern = "\\b${word}\\b".toRegex()
//    return pattern.findAll(body).count()
//}
//
//
//
//fun main() = with(System.`in`.bufferedReader()) {
//    val word = "Muzi"
//    val solution = Solution()
//    val pages = arrayOf(
//        "<html lang=\"ko\" xml:lang=\"ko\" xmlns=\"http://www.w3.org/1999/xhtml\">\n<head>\n  <meta charset=\"utf-8\">\n  <meta property=\"og:url\" content=\"https://careers.kakao.com/interview/list\"/>\n</head>  \n<body>\n<a href=\"https://programmers.co.kr/learn/courses/4673\"></a>#!MuziMuzi!)jayg07con&&\n\n</body>\n</html>", "<html lang=\"ko\" xml:lang=\"ko\" xmlns=\"http://www.w3.org/1999/xhtml\">\n<head>\n  <meta charset=\"utf-8\">\n  <meta property=\"og:url\" content=\"https://www.kakaocorp.com\"/>\n</head>  \n<body>\ncon%\tmuzI92apeach&2<a href=\"https://hashcode.co.kr/tos\"></a>\n\n\t^\n</body>\n</html>"
//
//    )
//    val result = solution.solution(word, pages)
//    println(result)
//
//}
//
//
//
//class Solution {
//    fun solution(word: String, pages: Array<String>): Int {
//        var answer = 0
//        val smallWord = word.lowercase()
//
//        val parsePages = pages.mapIndexed { index, page ->
//            val url = findUrl(page)
//            val links = findLinks(page)
//            val basicScore = wordsCount(page, smallWord)
//            Pages(index,url,basicScore,links)
//        }
//
//        val urlToPages = parsePages.associateBy { it.url }
//
//        parsePages.forEach {pages ->
//            pages.link.forEach {link ->
//                if (pages.link.isNotEmpty()) {
//                    urlToPages[link]?.matchingScore =
//                        (urlToPages[link]?.matchingScore ?: 0.0) + (pages.basicScore / pages.link.size.toDouble())
//                }
//            }
//        }
//
//        parsePages.forEach {page ->
//            page.totalScore = page.matchingScore + page.basicScore
//        }
//
//        answer = parsePages.maxByOrNull { it.totalScore }!!.index
//
//        return answer
//    }
//
//}
